keps/sig-scheduling/2249-pod-affinity-namespace-selector/README.md

```diff
@@ -0,0 +1,663 @@
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+I am running a SaaS service, the workload for each customer is placed in a separate
+namespace. The workloads requires 1:1 pod to node placement. I want to use pod 
+anti-affinity across all customers namespaces to achieve that.
+
+<!--
+
+### Notes/Constraints/Caveats (Optional)
+
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+The main risk is performance, using namespace selector will make it easier for 
+users to specify affinity constraints across a large number of namespaces.
+
+The initial implementation of pod affinity/anti-affinity suffered from 
+performance challenges, however over the releases 1.17 - 1.20 we significantly
+improved that. 
+
+We currently have integration test and clusterloader benchmarks that evaluate the extreme
+cases of all pods having affinity/anti-affinity constraints to/against each other.
+Those benchmarks show that the scheduler is able to achieve maximum throughput 
+(i.e., the api-server qps limit).
+
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+ Add `NamespaceSelector` field to `PodAffinityTerm`:
+
+```go
+type PodAffinityTerm struct {
+    // A label query over the set of namespaces for which the term applies to.
+    // The term is applied to the union of the namespaces selected by this field
+    // and the ones listed in the `Namespaces` field.
+	NamespaceSelector *metav1.LabelSelector
+}
+```
+
+As indicated in the comment, the scheduler will consider the union of the namespaces
+specified in the existing `Namespaces` field and the ones selected by the new 
+`NamespaceSelector` field. `NamespaceSelector`  is ignored when set to nil.
+
+The feature will be guarded by a new feature flag. When disabled, the api-server 
+will silently drop the `NamespaceSelector` field if set, moreover kube-scheduler 
+will ignore the field and continue to behave as before.
+
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Test Plan
+
```

Comment:
Explicitly call out correctness tests for namespace addition/removal/relabel scenarios. The scheduler's node caching has produced correctness problems in the past that we want to avoid here.

Reply:
done.